examples/sas_rec.py

```diff
@@ -0,0 +1,657 @@
+            ops.multiply(positive_sequence_embedding, item_sequence_embedding),
+            axis=-1,
+        )
+        negative_logits = ops.sum(
+            ops.multiply(negative_sequence_embedding, item_sequence_embedding),
+            axis=-1,
+        )
+        logits = ops.concatenate([positive_logits, negative_logits], axis=1)
+
+        # Labels
+        labels = ops.concatenate(
+            [
+                ops.ones_like(positive_logits),
+                ops.zeros_like(negative_logits),
+            ],
+            axis=1,
+        )
+
+        # sample weights
+        sample_weight = ops.concatenate(
+            [sample_weight, sample_weight],
+            axis=1,
+        )
+
+        loss = self.loss_fn(
+            y_true=ops.expand_dims(labels, axis=-1),
+            y_pred=ops.expand_dims(logits, axis=-1),
+            sample_weight=sample_weight,
+        )
+        loss = ops.divide_no_nan(ops.sum(loss), ops.sum(sample_weight))
+
+        return loss
+
+    def compute_output_shape(self, inputs_shape):
+        return list(inputs_shape) + [self.hidden_dim]
+
+
+"""
+Let's instantiate our model and do some sanity checks.
+"""
+
+model = SasRec(
+    vocabulary_size=movies_count + 1,
+    num_layers=NUM_LAYERS,
+    num_heads=NUM_HEADS,
+    hidden_dim=HIDDEN_DIM,
+    dropout=DROPOUT,
+    max_sequence_length=MAX_CONTEXT_LENGTH,
+)
+
+# Training
+output = model(
+    inputs={
+        "item_ids": ops.ones((2, MAX_CONTEXT_LENGTH), dtype="int32"),
+        "padding_mask": ops.ones((2, MAX_CONTEXT_LENGTH), dtype="bool"),
+    },
+    training=True,
+)
+print(output["item_sequence_embedding"].shape)
+
+# Inference
+output = model(
+    inputs={
+        "item_ids": ops.ones((2, MAX_CONTEXT_LENGTH), dtype="int32"),
+        "padding_mask": ops.ones((2, MAX_CONTEXT_LENGTH), dtype="bool"),
+    },
+    training=False,
+)
+print(output["predictions"].shape)
+
+"""
+Now, let's compile and train our model.
+"""
+
+model.compile(
+    optimizer=keras.optimizers.Adam(learning_rate=LEARNING_RATE, beta_2=0.98),
+)
+model.fit(
+    x=train_ds,
+    validation_data=val_ds,
+    epochs=NUM_EPOCHS,
+)
+
+"""
+## Making predictions
+
+Now that we have a model, we would like to be able to make predictions.
+
+So far, we have only handled movies by id. Now is the time to create a mapping
+keyed by movie IDs to be able to surface the titles.
+"""
+
+movie_id_to_movie_title = dict(zip(movies_df["MovieID"], movies_df["Title"]))
+movie_id_to_movie_title[0] = ""  # Because id 0 is not in the dataset.
+
+"""
+We then simply use the Keras `model.predict()` method. Under the hood, it calls
+the `BruteForceRetrieval` layer to perform the actual retrieval.
+
+Note that this model can retrieve movies already watched by the user. We could
+easily add logic to remove them if that is desirable.
+"""
+
+for ele in val_ds.unbatch().take(1):
+    test_sample = ele[0]
+    test_sample["item_ids"] = test_sample["item_ids"][tf.newaxis, ...]
```

Comment:
Wait, what is this doing? Why do you need `tf.newaxis`? Normally you use this in the context where you're adding dimensions. Is this doing `expand_dims(..., axis=0)`?

Reply:
Whoops, meant to use `ops.expand_dims`. Need to add a batch dimension for the forward pass